File: backend/src/main/java/com/increff/project/controller/ProductController.java
package com.increff.project.controller;

import com.increff.project.model.data.ProductData;
import com.increff.project.dto.ProductDTO;
import com.increff.project.model.form.ProductFilterForm;
import com.increff.project.model.form.ProductForm;
import com.increff.project.model.form.TsvForm;
import com.increff.project.model.helpers.PageSizeHelper;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/products")
@CrossOrigin(origins = "http://localhost:3000")
public class ProductController {

   @Autowired
   private  ProductDTO productDTO;


   /* -------- CREATE -------- */
   @RequestMapping(method = RequestMethod.POST)
   @ResponseStatus(HttpStatus.CREATED)
   public void addProduct(@Valid @RequestBody ProductForm form) {
      productDTO.addProduct(form);
   }

   /* -------- UPDATE -------- */
   // Note: Kept as POST because your original code used @PostMapping("/update")
   @RequestMapping(method = RequestMethod.PUT, path = "/update")
   @ResponseStatus(HttpStatus.ACCEPTED)
   public void updateProduct(@Valid @RequestBody ProductForm form) {
      productDTO.updateProduct(form);
   }

   /* -------- UPLOAD -------- */
   @RequestMapping(method = RequestMethod.POST, path = "/uploadForm")
   public Map<String, String> addMultipleProducts(@RequestBody TsvForm form)
   {
     return productDTO.addMultipleProduct(form);
   }

   /* -------- READ (PAGINATED) -------- */
   @RequestMapping(method = RequestMethod.POST,path = "/pages")
   public List<ProductData> getProducts(@RequestBody PageSizeHelper form) {
      return productDTO.getAllPaginated(form);
   }

   @RequestMapping(method = RequestMethod.POST,path = "/barcode")
   public ProductData getByBarcode(@RequestBody ProductFilterForm form) {
      return productDTO.findByBarcode(form);
   }

   @RequestMapping(method = RequestMethod.POST,path = "/client")
   public List<ProductData> getAllClientProducts(@RequestBody ProductFilterForm form) {
      return productDTO.getAllClientProducts(form);
   }


   /* -------- READ ALL -------- */
   @RequestMapping(method = RequestMethod.GET, path = "/allProducts")
   public List<ProductData> getAllProducts() {
      return productDTO.getAll();
   }
}

File: Frontend/app/controllers/productController.ts
import api from "../interceptor/axios";

/* ---------------- GET PAGINATED PRODUCTS ---------------- */
export function getProducts(page: number, size: number) {
  return api.post("/products/pages", { page, size });
}

/* ---------------- FILTER PRODUCTS ---------------- */
export function filterProducts(payload: {
  filter: "barcode" | "clientName";
  search: string;
  page?: number;
  size?: number;
}) {

  console.log("Filtering products with payload:", payload);

  // 🔴 BARCODE SEARCH
  if (payload.filter === "barcode") {
    return api.post("/products/barcode", {
      barcode: payload.search,
    });
  }

  // 🔴 CLIENT SEARCH (PAGINATED)
  return api.post("/products/clientName", {
    clientName: payload.search,
    page: payload.page ?? 0,
    size: payload.size ?? 10,
  });
}


/* ---------------- ADD PRODUCT ---------------- */
export function addProduct(payload: any) {
  return api.post("/products", payload);
}

/* ---------------- UPDATE PRODUCT ---------------- */
export function updateProduct(payload: any) {
  return api.put("/products", payload);
}

/* ---------------- TSV UPLOAD ---------------- */
export function uploadProductTSV(payload: { base64: string }) {
  console.log("Uploading TSV with payload:", payload);
  return api.post("/products/uploadForm",  {
    base64file: payload.base64
  });
}

